pkgs/ffigen/test/native_objc_test/protocol_test.dart

```diff
@@ -0,0 +1,199 @@
+// Copyright (c) 2024, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// Objective C support is only available on mac.
+@TestOn('mac-os')
+
+import 'dart:async';
+import 'dart:ffi';
+import 'dart:io';
+
+import 'package:objective_c/objective_c.dart';
+import 'package:test/test.dart';
+
+import '../test_utils.dart';
+import 'protocol_bindings.dart';
+import 'util.dart';
+
+void main() {
+  group('protocol', () {
+    setUpAll(() {
+      logWarnings();
+      // TODO(https://github.com/dart-lang/native/issues/1068): Remove this.
+      DynamicLibrary.open('../objective_c/test/objective_c.dylib');
+      final dylib = File('test/native_objc_test/protocol_test.dylib');
+      verifySetupFile(dylib);
+      DynamicLibrary.open(dylib.absolute.path);
+      generateBindingsForCoverage('protocol');
+    });
+
+    group('ObjC implementation', () {
+      test('Method implementation', () {
+        final protoImpl = ObjCProtocolImpl.new1();
+        final consumer = ProtocolConsumer.new1();
+
+        // Required instance method.
+        final result = consumer.callInstanceMethod_(protoImpl);
+        expect(result.toString(), 'ObjCProtocolImpl: Hello from ObjC: 3.14');
+
+        // Optional instance method.
+        final intResult = consumer.callOptionalMethod_(protoImpl);
+        expect(intResult, 579);
+
+        // Required instance method from secondary protocol.
+        final otherIntResult = consumer.callOtherMethod_(protoImpl);
+        expect(otherIntResult, 10);
+      });
+
+      test('Unimplemented method', () {
+        final protoImpl = ObjCProtocolImplMissingMethod.new1();
+        final consumer = ProtocolConsumer.new1();
+
+        // Optional instance method, not implemented.
+        final intResult = consumer.callOptionalMethod_(protoImpl);
+        expect(intResult, -999);
+      });
+    });
+
+    group('Manual DartProxy implementation', () {
+      test('Method implementation', () {
+        final proxyBuilder = DartProxyBuilder.new1();
+        final consumer = ProtocolConsumer.new1();
+        final proto = getProtocol('MyProtocol');
+        final secondProto = getProtocol('SecondaryProtocol');
+
+        final sel = registerName('instanceMethod:withDouble:');
+        final signature = getProtocolMethodSignature(proto, sel,
+            isRequired: true, isInstance: true);
+        final block = DartInstanceMethodBlock.fromFunction(
+            (Pointer<Void> p, NSString s, double x) {
+          return 'DartProxy: $s: $x'.toNSString();
```

Comment:
Use `p`? Or what is the `p` for?

Reply:
It's the method selector. I'll hide it in the code gen layer. The only way to hide it from this layer would be option 4 from the doc, which has technical issues.